Document how to Update Data Builder Docs (#605)
opensafely/documentation: docs/updating-the-docs.md

Make sure you commit both the changes to the `examples` and `databuilder/snippets` directories.

If you are updating multiple examples it might be easier to the extraction step happen each time you change the python modules, you can do this with `just databuilder/watch-examples`.

> ## Updating Data Builder Backend and Contract documentation

Reviewer: Nitpick: Titles in this section seem not to capitalize the first letter of each word. _Data Builder_ is an exception, because it's a proper noun (cf. common noun). By this reasoning, I think that _Backend_ and _Contract_ should be _backend_ and _contract_.

> !!! note "You will need both Python 3.7 and 3.10 on your PATH to make these updates"

Reviewer 2: suggestion:

* Maybe clarify briefly why you need both (linking to #493 might suffice)
* Maybe worth warning that this isn't supported for Windows as yet, if that's the case still. [^1]

[^1]: We could try Gitpod, but not sure how this would work out.

Reviewer 2: The Justfile seems bash-heavy. My assumption is that trying to run the same commands in Windows will fail.

> Some of the Data Builder documentation is generated from parts of databuilder's codebase.

Reviewer 2: thought: I suspect there's more effort on the reader's part when trying to work out whether "Data Builder" and "databuilder" are the same or different here. My opinion rather than fact, though.

I think I'd always favour "Data Builder" unless explicitly referring to actually code use — that is, `import databuilder`. For the case of the repository, I'd probably write "Data Builder repository" and link to the repository. Anyone reading, feel free to suggest counter-examples.

(We may have inconsistencies elsewhere and we can tidy these up, before we make the documentation more public.)

> Some of the Data Builder documentation is generated from parts of databuilder's codebase.

Reviewer 2: ```suggestion
Some of the Data Builder information in this OpenSAFELY documentation is generated from the [Data Builder source code](https://github.com/opensafely-core/databuilder).
```

There are three components used to achieve this:

> * [Data Builder](https://github.com/opensafely-core/databuilder/)

Reviewer: We refer to this as "the Data Builder repo" below. Could we do the same here? It indicates what "Data Builder" means in this context; that is, the repository.

Author: I've not mentioned the repo here because it's not the repo per se, but the tool itself that is the component.  I linked to the repos for each component because that's where the one can make changes.  I don't feel strongly about it though, do you think it's worth changing still?

Reviewer: That's reasonable. At some point, it would be useful to note down what we're referring to with Data Builder, databuilder, etc. I don't feel strongly about it, though 🙂

Author: The [canonical name is `Data Builder`](https://app.shortcut.com/ebm-datalab/epic/372/rename-to-data-builder?vc_group_by=day&ct_workflow=all&cf_workflow=500000005), but I suspect many of us are referring to `databuilder` accidentally or to mean the repo without thinking about it!

Reviewer 2: thought: I suspect there's more effort on the reader's part when trying to work out whether "Data Builder" and "databuilder" are the same or different here. My opinion rather than fact, though.

I think I'd always favour "Data Builder" unless explicitly referring to actually code use — that is, `import databuilder`. For the case of the repository, I'd probably write "Data Builder repository" and link to the repository. Anyone reading, feel free to suggest counter-examples.

(We may have inconsistencies elsewhere and we can tidy these up, before we make the documentation more public.)

* [A mkdocs plugin](https://github.com/opensafely-core/mkdocs-opensafely-databuilder/)
* [This documentation site](https://github.com/opensafely/documentation/)

> To update the [Backend](data-backends.md) or [Contracts](contracts-reference.md) pages:

Reviewer: Same nitpick as before 🙂

1. Make your changes in the [Data Builder repo](https://github.com/opensafely-core/databuilder/).  A new version of databuilder will be released once your changes are merged to `main`.

> 1. In the [documentation repo](https://github.com/opensafely/documentation/) run `just databuilder/upgrade-databuilder` to update the version of databuilder installed in the subdirectory's virtualenv.

Reviewer: I think we should emphasize that this is done after a new version of databuilder is released. (Do all merges to `main` prompt a new release?)

Author: > Do all merges to main prompt a new release?

Yep!  The `build_and_deploy` workflow tags the repo based on the semantic commit used, and builds a new docker image.  Technically there's a race condition where you could hit merge, and then run `just databuilder/upgrade-databuilder` before the tagging flow finishes, but I suspect it's very unlikely to happen and I'm not sure it would be a valuable addition to the docs?

Reviewer: I think I've misunderstood. The `main` workflow's `tag-new-version` step doesn't always create a new release (only for `fix`, `perf`, `feat`, and `BREAKING CHANGE`). If I merged a `docs` change to `main`, such as changing the docstring for `PatientDemographics`, then a new release wouldn't be created, so this step would be a noop, wouldn't it?

Author: TIL `docs` doesn't create a new release… arg.

I'll amend to make this part conditional, thanks for catching this!

Reviewer: Thank you 🙂

Reviewer: `github-tag-action` never fails to be confusing.

Author: Oh wait, we actually need a new version of databuilder to be built so we can use the tooling here because we only install a versioned build of it 😩

The alternative is to run the command with a local copy of the databuilder and then copy `public_docs.json` over, but I think that's going to make the documentation here very complicated.  However, the alternative feels like a foot gun waiting to happen 🤔

Reviewer 2: comment: I think this still need addressing :arrow_up:

Reviewer 2: suggestion: The quick fix might just be to say:
* a current limitation is that you need to wait for a new release before rebuilding the generated content
* note that releases are only made for certain pull requests

Make sure you commit both the changes to the `examples` and `databuilder/snippets` directories.

If you are updating multiple examples it might be easier to the extraction step happen each time you change the python modules, you can do this with `just databuilder/watch-examples`.

## Updating Data Builder backend and contract documentation

!!! note "You will need both Python 3.7 and 3.10 on your PATH to make these updates"

> !!! note "This workflow is not supported on Windows, it might work but we have not been able to test it"

Reviewer 2: ```suggestion
!!! note "This workflow is not supported, and has not been tested, on Windows."
```
